nexus/db-queries/src/db/queries/next_item.rs

```diff
@@ -584,13 +603,328 @@ impl<Item> ShiftGenerator<Item> for DefaultShiftGenerator<Item> {
     }
 }
 
+/// Select the next available item from a table, using a self-join.
+///
+/// This query is similar in spirit to the above `NextItem` query. However, it's
+/// implementation is different, and specifically designed to limit memory
+/// consumption of the overall query.
+///
+/// CockroachDB eagerly evaluates subqueries. This leads to high memory usage
+/// for the basic `NextItem` query, since that relies on a subquery that uses
+/// `generate_series()` to create the list of all possible items. That entire
+/// list must be materialized and held in memory.
+///
+/// In contrast, this query is implemented using a self-join between the
+/// existing table to be searched and the "next entry" (e.g., the value of the
+/// target column plus 1). This can result in lower memory consumption, since
+/// that series need not be stored in memory. Note that this relies on an index
+/// on the item column, which may be partial.
+///
+/// The full query looks like this:
+///
+/// ```sql
+/// SELECT IF(
+///     -- This condition detects if the table is empty.
+///     EXISTS(
+///         SELECT
+///             1
+///         FROM
+///             <table>
+///         WHERE
+///             time_deleted IS NULL
```

Comment:
Aren't we also checking the scope here?

In the code below, I see: `WHERE ScopeColumn::NAME = scope_key`

Reply:
Yes, added that in 75311d5a7